biz.aQute.bndlib/src/aQute/bnd/cdi/CDIAnnotations.java

```diff
@@ -304,4 +304,28 @@ private Document readXMLResource(Resource resource) {
 		}
 	}
 
+	static class Tuple {
+		final String	resourcePath;
+		final Resource	resource;
+		final Discover	discover;
+
+		public Tuple(String resourcePath, Resource resource, Discover discover) {
+			this.resourcePath = resourcePath;
+			this.resource = resource;
+			this.discover = discover;
+		}
+
+		boolean sameOrigin(Clazz c) {
+			if (!resource.getClass()
```

Comment:
It seems kind of kludgy to check the resource objects to be of the same type. I don't see that it does much since you use Jar.fromResource just like Analyzer.analyzeBundleClassPath which calls analyzeJar which creates the Clazz objects and puts them in classspace. I guess I am not sure under what conditions they would not be the same type?

Reply:
You have a hard problem given the current information held by Analyzer. Perhaps the answer here is a change to Analyzer so that it maintains information on the BCP entry from which a Clazz object comes. Analyzer.analyzeBundleClassPath could be updated to pass the BCP entry information to analyzeJar which could update some new data structure when it creates a new Clazz object. Then the code here could simply consult the data structure to determine which BCP entry owns a Clazz object (which is really what you need to know here.)

Your current attempts to reverse engineer this information are merely guess and not reliable.

Reply:
.. yeah. honestly I was thinking the same :(